otherlibs/unix/execvp.c

```diff
@@ -36,32 +37,126 @@ CAMLprim value unix_execvp(value path, value args)
                                     /* from smart compilers */
 }
 
-#ifdef HAS_EXECVPE
+#ifndef HAS_EXECVPE
+int unix_execvpe_emulation(const char * name,
+                           char * const argv[],
+                           char * const envp[]);
+#endif
 
 CAMLprim value unix_execvpe(value path, value args, value env)
 {
   char_os ** argv;
   char_os ** envp;
   char_os * wpath;
+  int err;
   caml_unix_check_path(path, "execvpe");
   argv = cstringvect(args, "execvpe");
   envp = cstringvect(env, "execvpe");
   wpath = caml_stat_strdup_to_os(String_val(path));
+#ifdef HAS_EXECVPE
   (void) execvpe_os((const char_os *)wpath, EXECV_CAST argv, EXECV_CAST envp);
+  err = errno;
+#else
+  err = unix_execvpe_emulation(wpath, argv, envp);
+#endif
   caml_stat_free(wpath);
   cstringvect_free(argv);
   cstringvect_free(envp);
-  uerror("execvpe", path);
+  unix_error(err, "execvpe", path);
   return Val_unit;                  /* never reached, but suppress warnings */
                                     /* from smart compilers */
 }
 
-#else
+#ifndef HAS_EXECVPE
 
-CAMLprim value unix_execvpe(value path, value args, value env)
+static int unix_execve_script(const char * path,
+                              char * const argv[],
+                              char * const envp[])
+{
+  size_t argc, i;
+  char ** new_argv;
+
+  /* Try executing directly.  Will not return if it succeeds. */
+  execve(path, argv, envp);
+  if (errno != ENOEXEC) return errno;
```

Comment:
I compared this to deviations from the C `execve` here from the the OpenBSD [libc execve code](https://github.com/openbsd/src/blob/master/lib/libc/gen/exec.c#L201), and can't find any changes in errno responses.  The OpenBSD control flow is convoluted due to removing a retry loop for ETXTBUSY and not refactoring the code.

The only change in behaviour I can spot is that argv[0] is passed as `sh` in OpenBSD, while `/bin/sh` is invoked by the exece.  I don't think this changes the behaviour of any shells.